### PR TITLE
[BEAM-360] Some updates related to dynamic work rebalancing of custom sources.

### PR DESCRIPTION
**Pull request by chamikaramj**
Adds a class 'iobase.BoundedSourceSplit' to represent dynamic work rebalancing results of custom sources.

Updates Dataflow runner specific code (apiclient.py) to support dynamic work rebalancing custom sources.

Updates 'OffsetRangeTracker' so that the result of 'position_at_fraction()'' is a 'long' instead of a 'float'.

### CODE REVIEW

**Comment by chamikaramj** [comment]
R: @robertwb , can you take a look ?

File: sdks/python/apache_beam/io/range_trackers.py [hunk 1]
```diff
@@ -163,7 +163,7 @@ def position_at_fraction(self, fraction):
       raise Exception(
           'get_position_for_fraction_consumed is not applicable for an '
           'unbounded range')
-    return (math.ceil(self.start_position() + fraction * (
+    return long(math.ceil(self.start_position() + fraction * (
```
Comment (robertwb) [inline sdks/python/apache_beam/io/range_trackers.py]: int(...)

File: sdks/python/apache_beam/io/filebasedsource.py [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+        file_name, 'rb', 'application/octet-stream')
+
+  @staticmethod
+  def _estimate_sizes_in_parallel(file_names):
+
+    def _calculate_size_of_file(file_name):
+      f = fileio.ChannelFactory.open(
+          file_name, 'rb', 'application/octet-stream')
+      try:
+        f.seek(0, os.SEEK_END)
```
Comment (robertwb) [inline sdks/python/apache_beam/io/filebasedsource.py]: This seems really expensive; would it make sense to add a method to ChannelFactory that gets the size of a file instead?

File: sdks/python/apache_beam/io/iobase.py [hunk 3]
```diff
@@ -258,12 +258,23 @@ class DynamicSplitResult(object):
...
 class DynamicSplitResultWithPosition(DynamicSplitResult):
+  """"""
 
   def __init__(self, stop_position):
     assert isinstance(stop_position, ReaderPosition)
     self.stop_position = stop_position
 
 
+class BoundedSourceSplit(DynamicSplitResult):
+  """"""
```
Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: Docstring?

Comment (robertwb) [inline sdks/python/apache_beam/io/iobase.py]: What sources would return this? I thought all dynamic splits were going to be in terms of positions, which is future-proof with SDF.

File: sdks/python/apache_beam/io/range_trackers.py (hunk 1 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/range_trackers.py]: Done.

File: sdks/python/apache_beam/io/filebasedsource.py (hunk 2 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/filebasedsource.py]: Yeah, that's a better approach.

Added a method to 'ChannelFactory' and updated 'filebasedsource' to use that.

File: sdks/python/apache_beam/io/iobase.py (hunk 3 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/iobase.py]: Reverted these changes.

**Comment by chamikaramj** [comment]
@robertwb , can you PTAL.

**Comment by chamikaramj** [comment]
Squashed commits into a single commit and updated the commit message.

Please let me know if anything else needs to be updated.

File: sdks/python/apache_beam/io/gcsio.py [hunk 4]
```diff
@@ -236,6 +236,21 @@ def exists(self, path):
     except IOError:
       return False
 
+  @retry.with_exponential_backoff(
+      retry_filter=retry.retry_on_server_errors_and_timeout_filter)
+  def size(self, path):
```
Comment (robertwb) [inline sdks/python/apache_beam/io/gcsio.py]: Might be worth filing a ticket to do bulk file operations here (and elsewhere).

**Comment by robertwb** [comment]
I assume you've manually checked this against actual GCS, right? If so, LGTM.

**Comment by chamikaramj** [comment]
Thanks. Yes, manually tested against actual GCS.

File: sdks/python/apache_beam/io/gcsio.py (hunk 4 above)
Comment (chamikaramj) [inline sdks/python/apache_beam/io/gcsio.py]: Created https://issues.apache.org/jira/browse/BEAM-453 for this.

**Comment by chamikaramj** [comment]
@robertwb can you merge this ?

Thanks.

**Comment by chamikaramj** [comment]
Thanks Robert. Closing pull request.

**Pull request closed without merging**